File: C/qcirc.c
#include <stdio.h>
#include <stdlib.h>
struct circq
{
    int front;
    int maxnum;
    int *array;
};
int q[10], rear, front; //q is an array/queue of integers

void initialize()
{
    rear=0;
    front=-1;
}

int isoverflow()
{
    if (q[front]!=0 && q[rear]!=0) return 1;
    return 0;
}

int isunderflow()
{
    if (front==rear-1) return 1;
    return 0;
}

int enqueue(int z)
{
    if (isoverflow())
    {
        puts("Queue is full!");
        return 999;
    }
    if (q[rear]==0 && q[front]!=0)
    {
        q[rear]=q[front];
        q[front]=0;
    }
    q[rear]=z;
    rear+=1;
    return 1;
}

int dequeue()
{
    if (isunderflow()) 
    {
        puts("Queue is already empty!");
        return 999;
    }
    return q[front];
    front++;
}

void show()
{
    int i;
    if (isunderflow()) puts("Queue empty!");
    else
    {
        for (i=0; i<10; i++)
        {
            if (q[i]==0) break;
            else printf("%d\t",q[i]);
        }    
    puts(" ");
    }
}

int main()
{
    puts("Program to implement circular queue");
    initialize();
    char next='y';
    int choice, e;
    puts("Welcome to Queue Menu!");
    puts("The queue size is 10"); //Try to input queue size from user
    do{
        puts("Choices:");
        puts("1. Enqueue an element (except 0)");
        puts("2. Dequeue an element");
        puts("3. View the entire queue");
        puts("4. Exit");
        puts("Enter your choice:");
        scanf("%d", &choice);
        fflush(stdin);
        switch (choice)
            {
            case 1:
                puts("Enter element to enqueue");
                scanf("%d", &e);
                fflush(stdin);
                enqueue(e);
                break;
            
            case 2:
                dequeue();
                break;
            
            case 3:
                show();
                break;

            case 4:
                exit(0);
                break;

            default:
                puts("Invalid Choice!");
                break;
            }
        puts("Do you want to perform another operation? Enter y if yes");
        scanf("%c", &next);
    
    }while (next=='y' || next=='Y');
    return 0;
}

File: Python/qcirc.py
#Program to implement circular queues
class qcirc():

	def __init__(self, size):
		self.size = size
		self.queue = [None for i in range(size)] #initialize
		self.front = self.rear = -1

	def is_full(self):
		if ((self.rear + 1) % self.size == self.front): 
			return 1 

	def is_empty(self):
		if(self.front == -1): 
			return 1

	def enqueue(self, x):
		if (self.is_full()):
                    print("Queue is Full")

		if ((not self.is_full()) and self.front == -1):
			self.front = 0
			self.rear = 0
			self.queue[self.rear] = x

		elif (not self.is_full()):
			self.rear = (self.rear + 1) % self.size
			self.queue[self.rear] = x
			
	def dequeue(self):
		if (self.is_empty()):
			print("Queue is empty!")
			return

		if (self.front == self.rear): #iff one element only
			temp=self.queue[self.front]
			self.front = -1
			self.rear = -1
			return temp

		temp = self.queue[self.front]
		self.front = (self.front + 1) % self.size
		return temp

	def display(self):
		if self.is_empty():
			print("Queue empty!")

		elif (self.rear >= self.front):
			print("Elements in the circular queue are:", end = " ")
			for i in range(self.front, self.rear + 1):
				print(self.queue[i], end = " ")
			print ()
			if self.is_full(): print("Queue is full") 

		else:
			print ("Elements in Circular Queue are: ", end = " ")
			for i in range(self.front, self.size):
				print(self.queue[i], end = " ")
			for i in range(0, self.rear + 1):
				print(self.queue[i], end = " ")
			print ()
			if self.is_full(): print("Queue is full")

print("Welcome to the Circular Queue Program!")
num = int(input("How many elements to keep in the queue? (max 20) "))
q = qcirc(num)
ch = 'y'
while (ch == 'y' or ch == 'Y'):
	print("Choices")
	print("1. Enqueue an element")
	print("2. Dequeue an element")
	print("3. View the entire queue")
	print("4. Exit")
	choice = int(input("Enter your choice "))
	if (choice == 1):
		data = int(input("Enter the element to enqueue "))
		q.enqueue(data)
	if (choice == 2):
		if (q.is_empty()):
			pass
		else:
			print("Dequeued value ", q.dequeue())
	if (choice == 3):
		q.display()
	if (choice == 4):
		break
	if choice not in [i for i in range(5)]: 
		print("Invalid Choice!")
	ch = input("Do you want to perform another operation? ")
